File: js/webgpu/internal/logmath.mjs
/**
 * Log-space arithmetic utilities.
 *
 * All DP values are stored as log-probabilities. NEG_INF represents
 * log(0) = -Infinity. logaddexp(a,b) = log(exp(a)+exp(b)) is the
 * fundamental "plus" operation in the log-semiring.
 */

export const NEG_INF = -Infinity;

/**
 * log(exp(a) + exp(b)), numerically stable.
 * @param {number} a
 * @param {number} b
 * @returns {number}
 */
export function logaddexp(a, b) {
  if (a === NEG_INF) return b;
  if (b === NEG_INF) return a;
  const m = a > b ? a : b;
  return m + Math.log(Math.exp(a - m) + Math.exp(b - m));
}

/**
 * max(a, b) — the "plus" in the max-plus (Viterbi) semiring.
 * @param {number} a
 * @param {number} b
 * @returns {number}
 */
export function logmax(a, b) {
  return a > b ? a : b;
}

/**
 * Semiring abstraction for log-space DP.
 * @typedef {'logsumexp'|'maxplus'} SemiringType
 */

/**
 * @param {SemiringType} type
 * @returns {{ plus: function, reduce: function }}
 */
export function makeSemiring(type) {
  if (type === 'maxplus') {
    return {
      plus: logmax,
      reduce(arr) {
        let m = NEG_INF;
        for (let i = 0; i < arr.length; i++) {
          if (arr[i] > m) m = arr[i];
        }
        return m;
      },
      argreduce(arr) {
        let m = NEG_INF;
        let idx = 0;
        for (let i = 0; i < arr.length; i++) {
          if (arr[i] > m) { m = arr[i]; idx = i; }
        }
        return { value: m, index: idx };
      }
    };
  }
  // logsumexp (default)
  return {
    plus: logaddexp,
    reduce(arr) {
      let m = NEG_INF;
      for (let i = 0; i < arr.length; i++) {
        if (arr[i] > m) m = arr[i];
      }
      if (m === NEG_INF) return NEG_INF;
      let s = 0;
      for (let i = 0; i < arr.length; i++) {
        if (arr[i] !== NEG_INF) s += Math.exp(arr[i] - m);
      }
      return m + Math.log(s);
    },
    argreduce: null  // not meaningful for logsumexp
  };
}

/**
 * Log-semiring matrix multiply: C[i][k] = plus_j(A[i][j] + B[j][k])
 * All matrices are flat Float64Array of size S*S, row-major.
 * @param {Float64Array} A - (S, S) row-major
 * @param {Float64Array} B - (S, S) row-major
 * @param {number} S - state count
 * @param {function} reduce - semiring reduce over array
 * @returns {Float64Array} C - (S, S) row-major
 */
export function matMul(A, B, S, reduce) {
  const C = new Float64Array(S * S);
  const tmp = new Float64Array(S);
  for (let i = 0; i < S; i++) {
    for (let k = 0; k < S; k++) {
      for (let j = 0; j < S; j++) {
        tmp[j] = A[i * S + j] + B[j * S + k];
      }
      C[i * S + k] = reduce(tmp);
    }
  }
  return C;
}

/**
 * Log-space identity matrix (0 on diagonal, NEG_INF elsewhere).
 * @param {number} S
 * @returns {Float64Array}
 */
export function logIdentity(S) {
  const I = new Float64Array(S * S).fill(NEG_INF);
  for (let i = 0; i < S; i++) I[i * S + i] = 0;
  return I;
}

/**
 * Matrix-vector multiply: out[i] = reduce_j(M[i][j] + v[j])
 * @param {Float64Array} M - (S, S) row-major
 * @param {Float64Array} v - (S,)
 * @param {number} S
 * @param {function} reduce
 * @returns {Float64Array}
 */
export function matVecMul(M, v, S, reduce) {
  const out = new Float64Array(S);
  const tmp = new Float64Array(S);
  for (let i = 0; i < S; i++) {
    for (let j = 0; j < S; j++) {
      tmp[j] = M[i * S + j] + v[j];
    }
    out[i] = reduce(tmp);
  }
  return out;
}

/**
 * Transposed matrix-vector multiply: out[j] = reduce_i(M[i][j] + v[i])
 * Used in forward direction: out[dst] = reduce_src(prev[src] + trans[src][dst])
 * @param {Float64Array} M - (S, S) row-major, M[src][dst]
 * @param {Float64Array} v - (S,) source vector
 * @param {number} S
 * @param {function} reduce
 * @returns {Float64Array}
 */
export function vecMatMul(v, M, S, reduce) {
  const out = new Float64Array(S);
  const tmp = new Float64Array(S);
  for (let dst = 0; dst < S; dst++) {
    for (let src = 0; src < S; src++) {
      tmp[src] = v[src] + M[src * S + dst];
    }
    out[dst] = reduce(tmp);
  }
  return out;
}

File: rust/fused-plan7/src/logmath.rs
//! Log-space arithmetic utilities.
//!
//! Faithful port of `js/webgpu/internal/logmath.mjs`. All DP values are stored
//! as log-probabilities. `NEG_INF` represents `log(0) = -inf`. The "plus" of the
//! semiring is `logaddexp` (Forward) or `max` (Viterbi); the array "reduce" is
//! the corresponding fold over an iterator of values.
//!
//! Wasm-safe: no allocation, no std I/O, no threads.

/// log(0) = -infinity.
pub const NEG_INF: f64 = f64::NEG_INFINITY;

/// `log(exp(a) + exp(b))`, numerically stable. Matches `logaddexp` in
/// `logmath.mjs` exactly (including the `a === NEG_INF` / `b === NEG_INF`
/// short-circuits, which keep `NEG_INF` an identity element).
#[inline]
pub fn logaddexp(a: f64, b: f64) -> f64 {
    if a == NEG_INF {
        return b;
    }
    if b == NEG_INF {
        return a;
    }
    let m = if a > b { a } else { b };
    m + ((a - m).exp() + (b - m).exp()).ln()
}

/// `max(a, b)` — the "plus" of the max-plus (Viterbi) semiring.
#[inline]
pub fn logmax(a: f64, b: f64) -> f64 {
    if a > b {
        a
    } else {
        b
    }
}

/// Which semiring a DP runs in.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Semiring {
    /// Forward: plus = logaddexp, reduce = logsumexp.
    LogSumExp,
    /// Viterbi: plus = max, reduce = max.
    MaxPlus,
}

impl Semiring {
    /// Parse the string semiring names used by the JS reference
    /// (`"logsumexp"` / `"maxplus"`).
    pub fn parse_name(s: &str) -> Option<Semiring> {
        match s {
            "logsumexp" => Some(Semiring::LogSumExp),
            "maxplus" => Some(Semiring::MaxPlus),
            _ => None,
        }
    }

    /// Semiring "plus": combine two scalars.
    ///
    /// For `LogSumExp` this is `logaddexp`; for `MaxPlus` this is `max`.
    #[inline]
    pub fn plus(self, a: f64, b: f64) -> f64 {
        match self {
            Semiring::LogSumExp => logaddexp(a, b),
            Semiring::MaxPlus => logmax(a, b),
        }
    }

    /// Semiring "reduce" over an iterator of values.
    ///
    /// Mirrors the JS `reduce(arr)`:
    ///  - max-plus: the maximum (NEG_INF if empty).
    ///  - logsumexp: `m + log(sum exp(x_i - m))` over finite entries, where `m`
    ///    is the max; returns NEG_INF if the max is NEG_INF.
    ///
    /// IMPORTANT: this performs a single max scan then a single sum scan, in the
    /// same order the JS does, so the float64 result is bit-identical for a given
    /// iteration order. Callers must pass values in the same order as the JS
    /// (`src` ascending) to stay bit-identical.
    #[inline]
    pub fn reduce<I>(self, iter: I) -> f64
    where
        I: IntoIterator<Item = f64> + Clone,
    {
        match self {
            Semiring::MaxPlus => {
                let mut m = NEG_INF;
                for x in iter {
                    if x > m {
                        m = x;
                    }
                }
                m
            }
            Semiring::LogSumExp => {
                let mut m = NEG_INF;
                for x in iter.clone() {
                    if x > m {
                        m = x;
                    }
                }
                if m == NEG_INF {
                    return NEG_INF;
                }
                let mut s = 0.0_f64;
                for x in iter {
                    if x != NEG_INF {
                        s += (x - m).exp();
                    }
                }
                m + s.ln()
            }
        }
    }
}

/// `log(x)` with `log(0) = NEG_INF` (the JS `safeLog` / `strToProb`+log idiom).
#[inline]
pub fn safe_log(x: f64) -> f64 {
    if x > 0.0 {
        x.ln()
    } else {
        NEG_INF
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn neg_inf_is_identity() {
        assert_eq!(logaddexp(NEG_INF, -1.5), -1.5);
        assert_eq!(logaddexp(-1.5, NEG_INF), -1.5);
        assert_eq!(logaddexp(NEG_INF, NEG_INF), NEG_INF);
        assert_eq!(logmax(NEG_INF, -2.0), -2.0);
    }

    #[test]
    fn logaddexp_matches_definition() {
        // log(exp(a)+exp(b)) computed naively for small finite values.
        for &(a, b) in &[(-1.0_f64, -2.0_f64), (0.0, 0.0), (-10.0, -0.1), (3.0, -3.0)] {
            let naive = (a.exp() + b.exp()).ln();
            assert!((logaddexp(a, b) - naive).abs() < 1e-12, "a={a} b={b}");
        }
    }

    #[test]
    fn reduce_semirings() {
        let xs = [-1.0_f64, -3.0, NEG_INF, -2.0];
        assert_eq!(Semiring::MaxPlus.reduce(xs.iter().copied()), -1.0);
        let lse = Semiring::LogSumExp.reduce(xs.iter().copied());
        let naive = ((-1.0_f64).exp() + (-3.0_f64).exp() + (-2.0_f64).exp()).ln();
        assert!((lse - naive).abs() < 1e-12);
        // Empty / all-NEG_INF reduce to NEG_INF in both semirings.
        assert_eq!(Semiring::LogSumExp.reduce(std::iter::empty()), NEG_INF);
        assert_eq!(
            Semiring::LogSumExp.reduce([NEG_INF, NEG_INF].iter().copied()),
            NEG_INF
        );
        assert_eq!(Semiring::MaxPlus.reduce(std::iter::empty()), NEG_INF);
    }

    #[test]
    fn semiring_parse_name() {
        assert_eq!(Semiring::parse_name("logsumexp"), Some(Semiring::LogSumExp));
        assert_eq!(Semiring::parse_name("maxplus"), Some(Semiring::MaxPlus));
        assert_eq!(Semiring::parse_name("nope"), None);
    }
}
